File: MasteringKotlin/app/src/main/java/com/devlokos/masteringkotlin/chapters/chapter11/advanced/AdvancedGenerics.kt
package com.devlokos.masteringkotlin.chapters.chapter11.advanced

/**
 * Capítulo 11 — Genéricos.
 * Nivel AVANZADO: Stack genérico — push(), pop(), peek().
 */
fun main() {
    runAdvancedGenericsExamples()
}

/**
 * Stack genérico: LIFO. El parámetro de tipo T lo hace reutilizable para cualquier tipo.
 */
class Stack<T> {
    private val items = mutableListOf<T>()

    fun push(item: T) {
        items.add(item)
    }

    fun pop(): T? {
        return if (items.isEmpty()) null else items.removeAt(items.size - 1)
    }

    fun peek(): T? {
        return items.lastOrNull()
    }

    fun isEmpty(): Boolean = items.isEmpty()
    fun size(): Int = items.size
}

fun runAdvancedGenericsExamples() {
    println("========================================")
    println("       GENERIC STACK")
    println("========================================")

    val intStack = Stack<Int>()
    intStack.push(1)
    intStack.push(2)
    intStack.push(3)
    println("Stack size: ${intStack.size()}")
    println("Peek: ${intStack.peek()}")
    println("Pop: ${intStack.pop()}, ${intStack.pop()}, ${intStack.pop()}, ${intStack.pop()}")

    val stringStack = Stack<String>()
    stringStack.push("first")
    stringStack.push("second")
    println("Peek: ${stringStack.peek()}")
    println("Pop: ${stringStack.pop()}")
    println("========================================")
}

File: MasteringDart/lib/chapters/chapter11/advanced/advanced_generics.dart
// ignore_for_file: avoid_print
//
//  advanced_generics.dart
//  MasteringDart
//
//  Capítulo 11 — Genéricos.
//  Nivel AVANZADO: Stack genérico — push(), pop(), peek().
//
//  Ejecutar con: dart run lib/chapters/chapter11/advanced/advanced_generics.dart

/// Stack genérico: LIFO. El parámetro de tipo T lo hace reutilizable para cualquier tipo.
class Stack<T> {
  final List<T> _items = [];

  void push(T item) {
    _items.add(item);
  }

  T? pop() {
    if (_items.isEmpty) return null;
    return _items.removeLast();
  }

  T? peek() {
    if (_items.isEmpty) return null;
    return _items.last;
  }

  bool get isEmpty => _items.isEmpty;
  int get size => _items.length;
}

void runAdvancedGenericsExamples() {
  print('========================================');
  print('       GENERIC STACK');
  print('========================================');

  final intStack = Stack<int>();
  intStack.push(1);
  intStack.push(2);
  intStack.push(3);
  print('Stack size: ${intStack.size}');
  print('Peek: ${intStack.peek()}');
  print('Pop: ${intStack.pop()}, ${intStack.pop()}, ${intStack.pop()}, ${intStack.pop()}');

  final stringStack = Stack<String>();
  stringStack.push('first');
  stringStack.push('second');
  print('Peek: ${stringStack.peek()}');
  print('Pop: ${stringStack.pop()}');
  print('========================================');
}

void main() {
  runAdvancedGenericsExamples();
}
